### PR TITLE
[CDAP-744] Make router log less verbose for invalid uris

### PR DESCRIPTION
**Pull request by bdmogal**
Minor change to make router logs less verbose

Issue: https://issues.cask.co/browse/CDAP-744

### CODE REVIEW

File: src/main/java/co/cask/http/RequestRouter.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ private boolean handleRequest(HttpRequest httpRequest, Channel channel, ChannelH
 
   public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) {
 
-    LOG.error("Exception caught in channel processing.", e.getCause());
+    LOG.trace("Exception caught in channel processing.", e.getCause());
```
Comment (anew) [inline src/main/java/co/cask/http/RequestRouter.java]: seems that this applies to all exceptions? Including OOM or other serious internal errors. Do we really want to disable the logging for all of those?

I think the right fix is to change the method that interprets the URL (getDestinationMethod) to throw different exceptions in different cases. One for 400 bad requests, such as malformed URLs, one for requests that can't be routed (404 not found), and one for internal errors (500). In the last case, we should log an error and return 500, in the other cases, we should only log with Trace.

Comment (bdmogal) [inline src/main/java/co/cask/http/RequestRouter.java]: That's true we shouldn't. Let me think about it. Will update the PR.

**Comment by bdmogal** [comment]
Used the response code inside `HandlerException` to log at varying levels in the router. Log at `ERROR` level if the error is an internal server error, `TRACE` level if it is a user error (`BAD_REQUEST`, `NOT_FOUND` or `METHOD_NOT_ALLOWED`).

Testing:
1. Bring up CDAP with this version of netty-http.
2. The following requests do not result in an ERROR log: `/v2/booboo`, `/v2/ping -X PUT`. 
3. Threw an `IllegalStateException` from the `/v3/namespaces` API and verified that an ERROR log shows up.

**Comment by anew** [comment]
It looks much better. 

I have a slight doubt about the methodInfo... the RequestRouter is a long-lived object that gets reused across requests, right? So, is it always true that if the method was matched if methodInfo != null? It seems that a successful request leaves its methodInfo in place, and if the next request leads to an exception during method matching, then methodInfo is != null, even though we have not matched a method. 

Or am I misreading it?

**Comment by bdmogal** [comment]
I'm not completely familiar with netty-http or netty itself, however, it seems like as long as `exceptionCaught` method is called when the next request leads to an exception, the if condition for `methodInfo` will reset methodInfo to null if it is not null already [here](https://github.com/caskdata/netty-http/blob/develop/src/main/java/co/cask/http/RequestRouter.java#L143). So we should be ok?

However, in terms of exception handling, we seem to only throw `INTERNAL_SERVER_ERROR` if `methodInfo` is not null.

**Comment by anew** [comment]
I think RequestRouter is recreated for every new channel (new request) by the pipeline factory, so it should be fine.

**Comment by anew** [comment]
LGTM

**Pull request merged**